RFC: Binary Distribution Format

theory opened the pull request:
Add a new RFC describing the proposed trunk binary distribution format for PGXN packages. Inspired by Python wheel and pgt.dev, aiming to support binaries for every OS and architecture supported by PostgreSQL itself, as well as many versions of PostgreSQL.

[Previous discussion](https://github.com/orgs/pgxn/discussions/2).

theory committed:
RFC: Binary Distribution Format

Add a new RFC describing the proposed trunk binary distribution format
for PGXN packages. Inspired by Python wheel and pgt.dev, aiming to
support binaries for every OS and architecture supported by PostgreSQL
itself, as well as many versions of PostgreSQL.

pgguru commented on text/0002-binary-distribution-format.md:
```diff
@@ -0,0 +1,367 @@
+*   `package`: Package name, e.g. `pgmq`, `postgis`, `pgAdmin`, `pg_top`.
+*   `version`: Distribution version in [SemVer] format without build metadata,
+    e.g., `0.8.6` or `1.0.0-beta`.
+*   `pg`: Major version of Postgres the binary was built against, e.g.,
+    `pg15`, `pg16`.
```
Did we determine there was never a need for minor version-specific trunk builds?  Would this format support that, if we expanded the version to cover the major version as a whole, while allowing a minor version (perhaps the existence of a minor version would imply that version onward for minor versions)?

theory replied:
I'm basing this on the [proposed ABI and API guidance](https://justatheory.com/2024/06/abi-api-guidance/), which has this key statement:

> In general, extension code that compiles and works with a minor release should also compile and work with any other minor release of the same major version, past or future.

But I see no reason why it couldn't be expanded to support `pg16.3` or whatever if the need was there.

pgguru commented on text/0002-binary-distribution-format.md:
```diff
@@ -0,0 +1,367 @@
+2.  Split the name into two parts at the `+` sign. The left part is the
+    package name and [SemVer]. The right part is the platform specification.
+
+3.  For the left part, split on the right-most dash. If the string to the
+    right of the dash is a valid [SemVer], then the left part is the package
+    name. If the right string is not a valid [SemVer], try again at the second
+    right-most dash and check again. Continue until a valid SemVer is produced
+    or else fail.
```
Semver are required to have 3 parts, with any suffix after a dash (presumably not with another semver-like system, so `3.1.2-0` would be valid, but `3.1.2-1.2.3`?  This part feels like it has the potential for some ambiguity to me to be honest, but I trust you're more familiar with semver than I am at this point. :-)

theory replied:
Bah, no, of course, `3.1.2-1.2.3` is a valid Semver.

So perhaps it instead needs to be that distribution names cannot start with a digit or use a digit after a dash. That would simplify things: the version starts at the first digit after a dash. Of existing distributions, it looks like only `DBT-2` would be a problem:

```psql
distinct name from distributions where name LIKE '%-%';
           name           
──────────────────────────
 DBT-2
 E-Maj
 foreign-keycloak-wrapper
 hashtypes-aleksabl
 neo4j-fdw
 passwd-fdw
 pg-json
 Slony-I
 soundex-function
 soundex-operator
 soundex-type
 trunklet-format
(12 rows)
```

Note that this does not apply to the name of the extensions, apps, or anything else inside the distribution, just the name of the thing being distributed.

theory replied:
Okay, added the [Package Name](/pgxn/rfcs/pull/3/files#diff-3a6d7e2784cf9bc361c4255353765fc71bc523285341b9f77b6787223165a19cR138-R145) type to #3. I think that will solve the problem. I'll rewrite this bit.

pgguru commented on text/0002-binary-distribution-format.md:
```diff
@@ -0,0 +1,367 @@
+    Each must be plain text or Markdown-formatted. In the latter case, they
+    may use the extension `.md`.
+
+*   Trunk, being an installation format intended to install pre-compiled
+    binaries and supporting files, does not include a `Makefile`, `configure`
+    file or any other tool for building the package contents.
```
Definitely think something like this would actually be useful; some build script or other recipe to validate that the generated binaries match the build options, as well as metadata including how it was built so you could compare to (say) `pg_config` output to confirm specific built options match or don't.

theory replied:
Yeah, that's part of #2. :-)

theory replied:
Oh, this _is_ #2 🤦🏻. But anyway, that's the point of the `trunk.json` file.

theory committed:
Simplify package name and semver parsing

Made possible by RFC-3 disallowing digits after dashes in package names.

theory committed:
Simplify package name/version parsing

Made possible by forbidding dots (.) in Terms in the metatdata spec
(80702c3), making it impossible for a package name to include a semver.

jcflack commented on text/0002-binary-distribution-format.md:
```diff
@@ -0,0 +1,367 @@
+Without the trunk binary distribution format, it will be difficult to build
+and deliver cross-platform binary distribution of all the packages on PGXN.
+
+## Prior art
```
I'll have more substantive comments on things in a bit. But as long as there's a prior-art section ...

As near as I can tell, the Tembo `trunk` format demonstrated the "expand directory prefixes to `pg_config` directories" pattern [since March 2023][trunkpgc]. Also, PL/Java demonstrated the same pattern [since January 2016][pljpgc]. ;)

[trunkpgc]: https://github.com/tembo-io/trunk/commit/f2d4ea402ba6926c3390a68dbc91301b7a130f56
[pljpgc]: https://github.com/tada/pljava/commit/34773e81c5bb2add2ec273582549c0c10d20e5f3

theory replied:
Well yes, that's part of why it's called the "Trunk" format. But you're right it should be mentioned here. I wasn't aware of PL/Java using, it, though.

jcflack replied:
Sure, I assumed the reason you hadn't mentioned the earlier PL/Java prior art was that you hadn't known about it, not that you deliberately snubbed it or anything. :)

jcflack commented on text/0002-binary-distribution-format.md:
```diff
@@ -0,0 +1,367 @@
+
+## Unresolved questions
+
+*    Should the archive format be Zip or tarball? PGXN had traditionally used
```
I can't simplify this choice, but maybe I can complicate it in a useful way.

This seems to be a recurring question for projects of all kinds, with very few actual technical arguments to be made either way. Both are widely supported. Either one can deliver the files. Maybe the most noticeable difference is that zip is more closely _associated in people's minds_ with Windows and tar is more _associated in people's minds_ with \*n\*x.

That association can sometimes be so strong it leads to problems: a quarter century ago, I was contributing to a project (I think it was the ANTLR parser generator) that actually offered downloads in _both_ formats. The choice of format had been overloaded with extra meaning: it was assumed that people using the tar download would want the text artifacts (sources and docs and such) to come out with \*n\*x-style line endings, and people using the zip download would want those same artifacts to come out with Windows-style line endings.

Tar, of course, doesn't have any notion of line-ending conversions. What you put in is what comes out. So the project built the tarballs from files with \*n\*xy line endings.

Zip _does_ have a notion of line-ending conversions, so you can build a zipfile from \*n\*xy text files and have them become Windowsy when extracted on a Windows box.

But __the normal way zip does that is irredeemably broken__.

For each entry in a zip, the extractor looks at a bit that flags that entry as a binary or text file. Line ending conversions are applied to text files only, and binary ones are of course left alone. Everything about that is ok; the problem is how that flag bit gets set in the first place. The zip file creator tool sets it by guessing. It samples some bytes of each file, and if it decides they look more ASCII/textish than binaryish, it sets the entry's "apparent ASCII/text file" bit to formally record the stupid guess it made. The extractor will then, deterministically, doing exactly as it's been told, corrupt during extraction every binary file that got mistakenly flagged as text. And ANTLR users were being hit by that.

If someone were to build a low-level zip-file creation tool that actually set the "apparent ASCII/text" flag explicitly according to the known text/binaryness of each file instead of guessing, then a zip created that way could be safely extracted anywhere. It's only the familiar ubiquitous creation tool that sets the flag by guessing. I wasn't able to build a better creation tool using the Java library API because that API's not low-level enough to let me control that flag.

It's tempting to say "well, tar then! it won't ever mess with any of the bytes!". But the line-ending issue hasn't gone away; even in 2025 it is still likely that whatever files in a distribution are source/doc/config/text/etc. will be friendlier to view and edit on the user's system if they get the right line endings. Which isn't really a technically difficult problem, as long as you don't try to do it the way zip did. (Thankfully, Mac OS fell in line with the \*n\*xy line endings, after Macs originally had yet a third different way for lines to end.)

For a format like trunk that wants to cryptographically hash entries, line-ending conversion makes it necessary to specify which form gets hashed. It should probably be the single known form actually stored in the archive. If extracted on a system with different line endings, it should be understood that extracted text files on the filesystem may hash to unexpected values even though the archive they came from verifies ok.

When I was later working on PL/Java, years after the ANTLR experience, the choice of format was made easy by noting that absolutely anyone looking at PL/Java is going to have Java installed, and Java's standard library has APIs for zip (and for jar, which is nothing but zip with a manifest file in a specified format that adds information on the content—so, a lot like trunk). If an extractor Java class is added to a jar and named as `Main-Class` in the manifest, the jar can be run with `java -jar foo.jar` and so be self-extracting in case a user has a stripped-down Java runtime without a CLI jar tool.

Because the Java API didn't let me get low-level enough to write zip files with the "apparent text" bits correctly non-guessingly set, I just used entries in the manifest to explicitly indicate which members should or shouldn't get text treatment. The regular CLI zip or jar tools would attach no meaning to those when extracting, but I could write the self-extractor class to do the right thing, and then just recommend `java -jar foo.jar` as always the right way to extract it. And from there it was a short step to also using `pg_config` to find the directories to extract into.

For trunk the choice is harder because you can't count on any single thing like Java being around. (Well, PostgreSQL! Maybe if there were a tarfile-fdw or zipfile-fdw and it was in core...)

Maybe there is a Rust (or Python, or some other) library with a zip API that would give explicit control over the "apparent text" flag when creating. You could create zip files using that and then extraction could safely rely on the flag. Of course you'd still have a custom extractor to do the `pg_config` directory expansion anyway, so it could also follow the PL/Java example and indicate per-entry text/binaryness in the manifest.

I hope that background's at least interesting, if the matter of zip's handling of text hadn't been explicitly raised before. There's [documentation on the extractor used in PL/Java][jrxdoc] (which is actually a few years [older than PL/Java][jrx1st]) going into more depth than I have gone into here.

[jrxdoc]: https://tada.github.io/pljava/pljava-packaging/apidocs/org/gjt/cuspy/JarX.html
[jrx1st]: https://github.com/tada/pljava/commit/92cd3f7247dcdc2f1d7d76b2b187de09fb4b4432

theory replied:
Wow, I was not aware of this whole line-ending thing. Interestingly, it looks like gitattributes supports [text file annotation](https://git-scm.com/docs/gitattributes#_text) similar to what you describe for jar files. As PGXN v1 uses zip files, I can see where this would be super useful. Sadly, it doesn't look like [get-archive uses this annotation](https://git-scm.com/docs/git-archive#ATTRIBUTES).
